### PR TITLE
first pass at fixing worker port selection issue

### PR DESCRIPTION
**Pull request by camicarballo**
Address issue #2320 and the port detection aspect of issue #2321 

Does a simple read of stdout before connecting to get the port the server is listening on and connects there.

### CODE REVIEW

File: work_queue/src/work_queue_server.c [hunk 1]
```diff
@@ -84,8 +84,6 @@ void mainloop(struct work_queue *queue, struct link *client)
 		ssize_t read = link_read(client, message, 1, time(NULL) + timeout);
```
Comment (dthain) [inline work_queue/src/work_queue_server.c:84]: Use `link_readline` here to do what you want.

File: work_queue/src/work_queue_server.c [hunk 2]
```diff
@@ -272,12 +270,26 @@ int main(int argc, char *argv[])
...
+
 	struct link *server_link = link_serve(server_port);
 	if(!server_link) {
 		printf("could not serve on port %d: %s\n", server_port,strerror(errno));
 		return 1;
 	}
 
+    // what is the port chosen for the server?
+    char addr[LINK_ADDRESS_MAX];
+    int local = link_address_local(server_link, &addr, &server_port);
```
Comment (dthain) [inline work_queue/src/work_queue_server.c:284]: The name of an array is the address of its first element.
So just pass `addr` not `&addr`

**Comment by dthain** [comment]
@camicarballo also fix this up to use `link_readline` instead of `link_read` one byte at a time.

**Comment by camicarballo** [comment]
It was originally a `link_readline` but changed to a `link_read` as you requested since we can't expect that any message will be followed by a newline.

**Comment by dthain** [comment]
Sorry, I think we are both mixed up here.

After looking again, I see that you are reading digits until you get to an opening brace,
and then assuming that it is the start of the JSON body.  That's non ideal b/c (1) Reading one byte at a time with `link_read` is highly inefficient at scale and (2) We may want the body to contain something other than JSON in the future.

Please set up the message envelope like this:

1 - Send an integer as a string followed by a newline.
2 - Send the message body (in this case JSON) immediately following.

That way, a message send looks roughly like this:

```
sprintf(line,"%d\n",length);
link_write(link,line,strlen(line));
link_write(link,data,length);
```

And a message receive looks roughly like this:

```
link_readline(link,line,sizeof(line));
link_read(link,data,length);
```

**Comment by dthain** [comment]
Looks good!

**Pull request merged**